### PR TITLE
Done

### PR DESCRIPTION
**Pull request by jasonpaulso**

### CODE REVIEW

File: bin/tictactoe [hunk 1]
```diff
@@ -1,3 +1,55 @@
...
+      game.play_cli
+    when '3'
+      game = Game.new(Player::Computer.new("X"), Player::Computer.new("O"))
+      game.play_cli
+    when "q"
+      exit
+    else
+      puts 'What was that? (You might have entered an invalid response.)'
+    end
+end
```
Comment (randallreedjr) [inline bin/tictactoe]: Prefer to encapsulate CLI logic into a class. For example, the only code in your bin executable should be:

```ruby
#!/usr/bin/env ruby
       
require_relative '../config/environment'

TicTacToeCLI.new.call
```

And then you'd have a file: `lib/tic_tac_toe_cli.rb` with:

```ruby
class TicTacToeCLI
  def initialize
  end

  def call
   # Your game logic
  end
end
```

The code in your bin would generally just go in `call`.

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
+require "pry"
+
+class Board
+  attr_accessor :cells
+
+  def initialize
+    @cells = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
```
Comment (randallreedjr) [inline lib/board.rb]: I would just call `self.reset!` here and delegate to that method to create the default empty board.

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def reset!
+    @cells.clear
+    @cells = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
```
Comment (randallreedjr) [inline lib/board.rb]: Another way to instantiate an Array with `i` values is `Array.new(9, " ")` (see [Array.new](http://ruby-doc.org/core-2.2.0/Array.html#method-c-new)), where `9` is the number of elements in the array and `" "` is the default value for the elements. It's more concise and will prevent errors from repeatedly typing `" "`.

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+class Board
+  attr_accessor :cells
+
+  def initialize
+    @cells = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
+  end
+
+  def reset!
+    @cells.clear
```
Comment (randallreedjr) [inline lib/board.rb]: No need to call clear as the contents of cells will be overwritten by the assignment.

File: lib/board.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    @cells.all? {|cell| cell == "X" || cell == "O"}
+  end
+  def turn_count
+    @cells.count{|token| token == "X" || token == "O"}
+  end
+  def taken?(position)
+    !(@cells[position.to_i - 1].nil? || @cells[position.to_i - 1] == " ")
+  end
+  def valid_move?(position)
+    return if (position.to_i - 1 < 0)
```
Comment (randallreedjr) [inline lib/board.rb]: You can simplify checking the valid integer range, and make it more semantic, by using [`Comparable#between?`](http://ruby-doc.org/core-2.2.3/Comparable.html#method-i-between-3F).
`position.between?(1,9)`

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    [6,4,2]
+  ]
+
+  def initialize(player_1=Player::Human.new("X"), player_2=Player::Human.new("O"), board=Board.new)
+    @board = board
+    @player_1 = player_1
+    @player_2 = player_2
+  end
+  def turn_count
+    self.board.cells.count{|token| token == "X" || token == "O"}
```
Comment (randallreedjr) [inline lib/game.rb]: Isn't this method already implemented as `self.board.turn_count`?

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  end
+  def current_player
+    self.turn_count % 2 == 0 ? player_1 : player_2
```
Comment (randallreedjr) [inline lib/game.rb]: Use [`Integer#even?`](http://ruby-doc.org/core-1.8.7/Integer.html#method-i-even-3F) or [`Integer#odd?`](http://ruby-doc.org/core-1.8.7/Integer.html#method-i-odd-3F)

`@board.turn_count.even? ? @player_1 : @player_2`

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    self.board.cells.all? {|cell| cell == "X" || cell == "O"}
+  end
+  def draw?
+    won?.nil? && full?
+  end
+  def over?
+    draw? || won? 
+  end
+  def winner
+    won?.each {|num| return self.board.cells[num][0]} unless won?.nil?
```
Comment (randallreedjr) [inline lib/game.rb]: Why use `each` and then `return` right away? I don't think an iterator is necessary here.
```
def winner
  self.board.cells[0][0] if won?
end
```

File: lib/players/human.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class Player::Human < Player
+  def move(board=nil)
+    puts "Hello, Player #{token}, what is your move(1-9)?"
+    move = gets.chomp
```
Comment (randallreedjr) [inline lib/players/human.rb]: You don't need to store your return value in a local variable; just utilize the implicit return value of the expression.
```
def move(board)
  puts "Where would you like to move?"
  gets.chomp
end
```

File: lib/game.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      position_3 = self.board.cells[win_index_3]
+
+      position_1 == "X" && position_2 == "X" && position_3 == "X" || position_1 == "O" && position_2 == "O" && position_3 =="O"
+    end
+  end
+  def full?
+    self.board.cells.all? {|cell| cell == "X" || cell == "O"}
+  end
+  def draw?
+    won?.nil? && full?
```
Comment (randallreedjr) [inline lib/game.rb]: You could also use `!won?`; having two methods with `?` looks a little funny.

**Comment by randallreedjr** [comment]
@jasonpaulso Great job! I left you a few notes, feel free to respond with any questions or comments. Sorry it took me so long to provide you with feedback, hopefully it’s still useful to you! Closing this PR for now.

**Comment by jasonpaulso** [comment]
Thanks Randall, this is very helpful!

**Pull request closed without merging**